### PR TITLE
Added schema generation for Generic fields

### LINKED ISSUE #1578: BaseModel.schema() fails when using generic types
# Bug
## Version
Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
  pydantic version: 1.5.1
            pydantic compiled: True
                 install path: C:\Users\adrian\miniconda3\envs\sangl-butler\Lib\site-packages\pydantic
               python version: 3.8.1 (default, Mar  2 2020, 13:06:26) [MSC v.1916 64 bit (AMD64)]
                     platform: Windows-10-10.0.18362-SP0
     optional deps. installed: ['typing-extensions', 'email-validator']

```
## Description
Hi there!

I want to write a generic type, that behaves like the type it extends, but injects fields into the model with ``__modify_schema__``. So my approach was to use a generic class to use it like that:

```py
from pydantic import BaseModel

class MyModel(BaseModel):
   my_field: FieldThatModifiesModel[str] = "Default value"

````
The field should completely behave like the type var.

However, when I use a generic type in my model, I can't retrieve the model anymore.
See the code below (taken from the docs):
## Example
```py
from pydantic import BaseModel, ValidationError
from pydantic.fields import ModelField
from typing import TypeVar, Generic

AgedType = TypeVar('AgedType')
QualityType = TypeVar('QualityType')

# This is not a pydantic model, it's an arbitrary generic class
class TastingModel(Generic[AgedType, QualityType]):
    def __init__(self, name: str, aged: AgedType, quality: QualityType):
        self.name = name
        self.aged = aged
        self.quality = quality

    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    # You don't need to add the "ModelField", but it will help your
    # editor give you completion and catch errors
    def validate(cls, v, field: ModelField):
        if not isinstance(v, cls):
            # The value is not even a TastingModel
            raise TypeError('Invalid value')
        if not field.sub_fields:
            # Generic parameters were not provided so we don't try to validate
            # them and just return the value as is
            return v
        aged_f = field.sub_fields[0]
        quality_f = field.sub_fields[1]
        errors = []
        # Here we don't need the validated value, but we want the errors
        valid_value, error = aged_f.validate(v.aged, {}, loc='aged')
        if error:
            errors.append(error)
        # Here we don't need the validated value, but we want the errors
        valid_value, error = quality_f.validate(v.quality, {}, loc='quality')
        if error:
            errors.append(error)
        if errors:
            raise ValidationError(errors, cls)
        # Validation passed without errors, return the same instance received
        return v

class Model(BaseModel):
    # for wine, "aged" is an int with years, "quality" is a float
    wine: TastingModel[int, float]
    # for cheese, "aged" is a bool, "quality" is a str
    cheese: TastingModel[bool, str]
    # for thing, "aged" is a Any, "quality" is Any
    thing: TastingModel

print(Model.schema())
"""
Traceback (most recent call last):
  File ".../generic_type.py", line 54, in <module>
    print(Model.schema())
  File "pydantic\main.py", line 556, in pydantic.main.BaseModel.schema
  File "pydantic\schema.py", line 132, in pydantic.schema.model_schema
  File "pydantic\schema.py", line 455, in pydantic.schema.model_process_schema
  File "pydantic\schema.py", line 491, in pydantic.schema.model_type_schema
  File "pydantic\schema.py", line 185, in pydantic.schema.field_schema
  File "pydantic\schema.py", line 411, in pydantic.schema.field_type_schema
AssertionError: 10
"""
```

### PR DESCRIPTION
**Pull request by maximberg**
## Change Summary

Generating schema for generic fields hasn't been implemented. Now it should work.

## Related issue number

Fixes #1578. 

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

**Comment by PrettyWood** [comment]
Hi @maximberg and thanks for the contribution.
I don't know if you saw but #2260 was opened yesterday and tackles the same issue.
Probably good to merge both works!

**Comment by maximberg** [comment]
Hi @PrettyWood, 
I didn't see your pull request yesterday. See what you can add to mine if it's not difficult for you. Tests passed, but coverage decreased.

**Comment by maximberg** [comment]
Coverage restored.

File: pydantic/schema.py [hunk 1]
```diff
@@ -496,7 +498,10 @@ def field_type_schema(
 
     # check field type to avoid repeated calls to the same __modify_schema__ method
     if field.type_ != field.outer_type_:
-        modify_schema = getattr(field.outer_type_, '__modify_schema__', None)
+        if field.shape == SHAPE_GENERIC:
+            modify_schema = getattr(field.type_, '__modify_schema__', None)
+        else:
+            modify_schema = getattr(field.outer_type_, '__modify_schema__', None)
```
Comment (Mazyod) [inline pydantic/schema.py:504]: suggestion for DRYness

```suggestion
        if field.shape == SHAPE_GENERIC:
            field_type = field.type_
        else:
            field_type = field.outer_type_
        modify_schema = getattr(field_type, '__modify_schema__', None)
```

**Comment by PrettyWood** [comment]
> I didn't see your pull request yesterday. See what you can add to mine if it's not difficult for you. Tests passed, but coverage decreased.

As a matter of fact it's not my PR but @danni's. I just think both works should be merged into one. If this PR is good to go and test cases from the @danni's PR are green with your PR fine! But I won't checkout locally both works and do it myself sorry

**Comment by maximberg** [comment]
@Mazyod Thanks! I used your proposal.

**Comment by maximberg** [comment]
@PrettyWood Actually @danni's PR not exactly about problem, described in #1578.

**Comment by Mazyod** [comment]
@samuelcolvin I wanted to sponsor this project because recent generics enhancements will make a big difference for our project 🙇‍♂️ Appreciate your help

**Comment by davidhyman** [comment]
:+1: here too

I have tried out both #2262 (this) and #2260 MRs and they both work well with my codebase :smile: 

Currently @danni's MR is indicating lower code coverage. However, two of those tests flag failures on @maximberg's MR (`test_nested_generic`, `test_complex_nested_generic`), and I see (but don't fully comprehend) the diff between them.

By way of expediency, perhaps this MR could be merged as-is, providing an improvement? A later MR would then add the nested generic functionality.

In the meantime I will attempt to understand both MRs...
- I could then try and bump coverage for #2260 ... maybe @danni has time to take a look?
- Alternatively, I suppose I could try and merge both... :scientist:

**Comment by maximberg** [comment]
@davidhyman My code solves the problem of generic fields in generating schema. 
@danni tries to solve the problem of nested generic models. 
In my project I want to use generic fields and I made necessary changes here.

**Comment by maximberg** [comment]
I found some strange behavior when generic field is Optional. Will check it Monday, 15th.

**Comment by samuelcolvin** [comment]
this looks great, thank you.

If you find other strange behaviour, please create an issue or if you're confident, a PR.

**Pull request merged**